### PR TITLE
Isothermal constraint

### LINKED ISSUE #649: Isothermal configuration arguments
Almost all models in WaterTAP should have an isothermal configuration argument. Where the energy balance on the control volume is skipped and instead the energy balance is just represented with making the outlet temperatures equal to the inlet temperature.

The only models this won't apply to are processes where temperature change is an inherent part of the process like evaporative processes and some chemical reactions.

### PR DESCRIPTION
**Pull request by bknueven**
## Fixes/Resolves: #649 

(replace this with the issue # fixed or resolved, if no issue exists then a brief statement of what this PR does)

## Changes proposed in this PR:
- Add custom CV with isothermal assumption
- Unit Models updated:
   - [x] Boron removal
   - [x] Coag and floc
   - [x] Electrodialysis 0D
   - [x] Electrodialysis 1D
   - [x] GAC
   - [x] Nanofiltration ZO
   - [x] Pressure Exchanger

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/unit_models/nanofiltration_ZO.py [hunk 1]
```diff
@@ -442,15 +464,6 @@ def eq_recovery_mass_phase_comp(b, t, j):
...
 
-        @self.Constraint(
-            self.flowsheet().config.time, doc="Isothermal assumption for permeate"
-        )
-        def eq_permeate_isothermal(b, t):
-            return (
-                b.feed_side.properties_in[t].temperature
-                == b.properties_permeate[t].temperature
-            )
-
```
Comment (bknueven) [inline watertap/unit_models/nanofiltration_ZO.py]: You don't want to remove this one -- it's connecting the feed and permeate sides. This change is just about connected the feed side inlet and outlet temperatures.

File: watertap/unit_models/nanofiltration_ZO.py [hunk 2]
```diff
@@ -678,9 +688,6 @@ def calculate_scaling_factors(self):
             sf = iscale.get_scaling_factor(self.rejection_phase_comp[ind])
             iscale.constraint_scaling_transform(c, sf)
 
-        for t, c in self.eq_permeate_isothermal.items():
-            sf = iscale.get_scaling_factor(self.feed_side.properties_in[t].temperature)
-            iscale.constraint_scaling_transform(c, sf)
```
Comment (bknueven) [inline watertap/unit_models/nanofiltration_ZO.py]: Please retain this (see above).

File: watertap/unit_models/gac.py [hunk 3]
```diff
@@ -244,16 +285,6 @@ def eq_isothermal(b, t):
...
-        @self.Constraint(
-            self.flowsheet().config.time,
-            doc="Isothermal assumption for absorbed contaminant",
-        )
-        def eq_isothermal_adsorbed_contam(b, t):
-            return (
-                b.process_flow.properties_in[t].temperature
-                == b.adsorbed_contam[t].temperature
-            )
-
```
Comment (bknueven) [inline watertap/unit_models/gac.py]: This should be retained.

**Review by bknueven — COMMENTED** [review]
@MarcusHolly there was a few places where the isothermal constraint was crossing the control volume. I've restored those and now I'm confident the test should pass.

File: watertap/core/control_volume_isothermal.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                b.properties[t, b.length_domain.first()].temperature
+                == b.properties[t, x].temperature
+            )
+
+    def calculate_scaling_factors(self):
+        super().calculate_scaling_factors()
+        if hasattr(self, "isothermal_assumption_eq"):
+            for (t, x), c in self.isothermal_assumption_eq.items():
+                sf = iscale.get_scaling_factor(
+                    self.properties[t, x].temperature, default=1e2, warning=True
```
Comment (lbibl) [inline watertap/core/control_volume_isothermal.py]: Is the default sf=1e2 for temperature too larger?

Comment (MarcusHolly) [reply watertap/core/control_volume_isothermal.py]: Oh, yea this should probably be 1e-2, right? Thanks for pointing that out.

Comment (bknueven) [reply watertap/core/control_volume_isothermal.py]: ~~Temperature is measured in Kelvins, and all our processes have to operate well above 200K. `1e2` seems appropriate?~~

Right, `1e-2`.

File: watertap/core/control_volume_isothermal.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        )
+        def isothermal_assumption_eq(b, t):
+            return b.properties_out[t].temperature == b.properties_in[t].temperature
+
+    def calculate_scaling_factors(self):
+        super().calculate_scaling_factors()
+        if hasattr(self, "isothermal_assumption_eq"):
+            for t, c in self.isothermal_assumption_eq.items():
+                sf = iscale.get_scaling_factor(
+                    self.properties_in[t].temperature, default=1e2, warning=True
```
Comment (bknueven) [inline watertap/core/control_volume_isothermal.py]:
```suggestion
                    self.properties_in[t].temperature, default=1e-2, warning=True
```

File: watertap/core/control_volume_isothermal.py (hunk 4 above)
Comment (bknueven) [inline watertap/core/control_volume_isothermal.py]:
```suggestion
                    self.properties[t, x].temperature, default=1e-2, warning=True
```

Comment (bknueven) [reply watertap/core/control_volume_isothermal.py]: Fixed in fb5a25d.

File: watertap/core/control_volume_isothermal.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            self.length_domain,
+            doc="Isothermal assumption for 1D control volume",
+        )
+        def isothermal_assumption_eq(b, t, x):
+            if x == b.length_domain.first():
+                return Constraint.Skip
+            return (
+                b.properties[t, b.length_domain.first()].temperature
+                == b.properties[t, x].temperature
+            )
```
Comment (TimBartholomew) [inline watertap/core/control_volume_isothermal.py:86]: I know this would be a separate PR, but creating this WaterTAP control volume has me thinking... What if we make a WaterTAP control volume that has a configuration argument for whether it is 0D or 1D and it automatically includes the references for the 0D so that it quacks like a 1D model with `properties[t, x]` instead of just `properties_in[t]` and `properties_out[t]`

File: watertap/unit_models/boron_removal.py [hunk 7]
```diff
@@ -120,23 +120,34 @@ class BoronRemovalData(InitializationMixin, UnitModelBlockData):
...
-    # NOTE: This option is temporarily disabled
-    '''
-    CONFIG.declare("energy_balance_type", ConfigValue(
-        default=EnergyBalanceType.useDefault,
-        domain=In(EnergyBalanceType),
-        description="Energy balance construction flag",
-        doc="""Indicates what type of energy balance should be constructed,
-    **default** - EnergyBalanceType.useDefault.
+    CONFIG.declare(
+        "isothermal",
```
Comment (TimBartholomew) [inline watertap/unit_models/boron_removal.py]: I think following IDAES conventions this might be better as `is_isothermal`

**Review by TimBartholomew — COMMENTED** [review]
2 comments - one just an idea for a separate effort, the other minor naming

File: watertap/core/control_volume_isothermal.py (hunk 6 above)
Comment (andrewlee94) [reply watertap/core/control_volume_isothermal.py:86]: My only caution here is to think through what this means downstream and what confusions this might create. IDAES deliberately kept the two separate and with different APIs so that it was clear to the end-user which was which, even if it meant they had to understand a bit more.

Comment (bknueven) [reply watertap/core/control_volume_isothermal.py:86]: I think there's a trade-off here between keeping separate, understandable APIs while also allowing code-sharing between 0D unit models and their 1D equivalent. I would also add that our 0D-model still allow you to access the 0D-type API.

I would be in favor of enabling a 1D-quacking 0DCV, with the caveat that for models for which we only have a 0D version of we stick to the 0D API in the implementation.

Comment (lbibl) [reply watertap/core/control_volume_isothermal.py:86]: Question: is "1D-quacking 0DCV" referring to a 1D-cv with 1 finite element?

Comment (bknueven) [reply watertap/core/control_volume_isothermal.py:86]: No -- it has (for a flow-forward model) `properties_in[t]` mapped to `properties[0,t]` and `properties_out[t]` mapped to `properties[1,t]` using references.

This allows for the expression of constraints common to both a 0D and 1D version of a model, see, e.g., https://github.com/watertap-org/watertap/blob/d2382933c88859b6e4f6ae7f791ca187f30a3a33/watertap/unit_models/reverse_osmosis_base.py#L346-L372

File: watertap/unit_models/boron_removal.py (hunk 7 above)
Comment (bknueven) [reply watertap/unit_models/boron_removal.py]: Done in 242f87c.

**Comment by TimBartholomew** [comment]
@bknueven the test failure is for the dye desalination case study, I think this came up before in another PR or was it always this PR? Do we have any insight into the issues here?

**Comment by bknueven** [comment]
> @bknueven the test failure is for the dye desalination case study, I think this came up before in another PR or was it always this PR? Do we have any insight into the issues here?

It's been an issue in this PR, and I think it's come up in others before.

I don't have any insight because I cannot reproduce the failure locally, even after creating a fresh conda environment on my Ubuntu VM. I would be happy if anybody else saw the same failure at the Linux GHA runners.

**Comment by bknueven** [comment]
> @bknueven the test failure is for the dye desalination case study, I think this came up before in another PR or was it always this PR? Do we have any insight into the issues here?

@TimBartholomew I managed to track this down. One of the steps in the dye desalination initialization converged for me locally, but only after 2000+ iterations. Fortunately, the fix was fairly simple: 8c8798e.

**Review by TimBartholomew — APPROVED** [review]
LGTM

**Comment by TimBartholomew** [comment]
@lbibl can you review/approve?

**Review by lbibl — APPROVED** [review]
Good to me.
